Fix zwave_js unique ID migration logic

raman325 opened the pull request:
## Proposed change
There are two things we want to fix from https://github.com/home-assistant/core/pull/46979:
1. The migration logic looked for a unique ID of `<home_id>.<old_value_id>` but we should have been looking for `<home_id>.<node_id>.<old_value_id>`. This is because the `unique_id` used `ZwaveDiscoveryInfo.value_id` which confusingly uses the `<node_id>.<value_id>` format. Because this logic was wrong, b0 users had entities duplicated.
2. In the entity instance, we were still using `ZwaveDiscoveryInfo.value_id`, so the unique ID was set to`<home_id>.<node_id>.<new_value_id>`. Even if the migration logic had worked, it would have still resulted in duplicates because we were migrating to `<home_id>.<new_value_id>` but the new entities are using this same `ZwaveDiscoveryInfo.value_id` format for unique ID except with the new value ID (`ZwaveValue.value_id`) format, so we would have successfully migrated the existing entities but they would still be orphaned.

`ZwaveDiscoveryInfo.value_id` is clearly a source of confusion (at least it was for me), it's only being used for unique IDs, and it doesn't add any uniqueness to the ID since node ID is already in the value ID, so I got rid of it entirely. I also updated the migration logic to look for both `<home_id>.<node_id>.<OLD_value_id>` (from 2021.2) and `<home_id>.<node_id>.<NEW_value_id>` from (2021.3.0b0) and 
to change them to `<home_id>.<NEW_value_id>`. This is less confusing as there is no duplicate information and the code is easier to read as a result.

Notification binary sensors have a unique ID that also includes the state key so there is additional logic added to handle that scenario.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

raman325 committed:
fix zwave_js unique ID migration logic

MartinHjelmare commented:
Many tests are failing.

raman325 committed:
fix mypy and add migration for b0 users

raman325 committed:
add support for notification binary sensor migration

raman325 committed:
fix test

raman325 commented:
When I wrote the original migration script, I did not try to actually perform the migration locally. That was a mistake. This time I ran master to load up entities, then switched to this dev branch and `zwave-js-server` 1.0.0 and verified that all the entities were migrated properly.

raman325 committed:
improve readability

raman325 committed:
improve debug comment

raman325 committed:
make code more readable

raman325 committed:
remove extra comments

raman325 committed:
get rid of unnecessary f-strings and fix paranthesis

balloob commented on homeassistant/components/zwave_js/__init__.py:
```diff
@@ -85,6 +85,21 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    @callback
+    def migrate_entity(platform: str, old_unique_id: str, new_unique_id: str) -> None:
+        """Check if entity with old unique ID exists, and if so migrate it to new ID."""
+        if entity_id := ent_reg.async_get_entity_id(platform, DOMAIN, old_unique_id):
```
Walrus!

balloob reviewed (approved):
This code looks good 👍

Nixon506E commented:
I made the mistake of updating all of my entities to the b0 migration and now when updating I lose all of my updated entities and no new entities are showing up. Can we get a migration from the b0 ids as well?

raman325 commented:
@Nixon506E that is covered, last thing I wanted to do was to burn you twice 😄

gregg098 commented:
I'm in the same boat as @Nixon506E. I migrated last night. I actually thought it was my fault in upgrading ZwaveJS2MQTT before the RC. So I spent a couple hours migrating entities over. Now I have a mix of unavailable and available. Downgrading seems to make things work. Going to roll back my backup to last night and see what happens. Not happy right now.

karl-gustav commented:
@raman325 so just to double check; we are not supposed to clean up the orphaned entities, those will ve fixed with the next beta release?

gregg098 commented:
Rolling back to 2021.3.0b0 and my ZwaveJS2MQTT docker data backup from last night worked, so I will stay there for awhile.

If it helps. Here is what I did last night.

I upgraded to 2021.3.0b0 and ZWaveJS2MQTT 1.4.0. I had double entities in every device for Zwave. I painfully deleted each unavailable entity to reclaim the entity_id, then renamed the new entity the same. Restarted Home Assistant again and cleaned up the remaining straggler entities by enabled EVERYTHING. Every single disable Zwave entity. Restarted Home Assistant again. This forced the old newly enabled entities to show up as Unavailable and allow me to delete them from the UI. I deleted all of these and re-disabled the stuff I didnt want;  random alarm level, access control, tamper sensors, etc. Some devices had 10-15 entities and I only wanted two of them. Restarted Home Assistant again and and everything was in good working order as of this morning. As far as I know, I cleaned out every single old entity. From what you're saying, it sounds like this should have worked, but clearly only partially worked. after upgrading to 2021.3.0b1.

Is there a fix? Or will I just have to bite the bullet eventually and re-do everything again if I want to get past 2021.3.0b0?

raman325 commented:
@karl-gustav only one set of entities will get migrated over because two entities can't have the same unique ID. If you leave orphaned entities there and upgrade, some set of entities will stay orphaned and some will be migrated over. There are reports on Discord that the update from b0 to b1 is causing all entities to show up as unavailable, which I am looking into.

raman325 commented:
> Rolling back to 2021.3.0b0 and my ZwaveJS2MQTT docker data backup from last night worked, so I will stay there for awhile.
> 
> If it helps. Here is what I did last night.
> 
> I upgraded to 2021.3.0b0 and ZWaveJS2MQTT 1.4.0. I had double entities in every device for Zwave. I painfully deleted each unavailable entity to reclaim the entity_id, then renamed the new entity the same. Restarted Home Assistant again and cleaned up the remaining straggler entities by enabled EVERYTHING. Every single disable Zwave entity. Restarted Home Assistant again. This forced the old newly enabled entities to show up as Unavailable and allow me to delete them from the UI. I deleted all of these and re-disabled the stuff I didnt want; random alarm level, access control, tamper sensors, etc. Some devices had 10-15 entities and I only wanted two of them. Restarted Home Assistant again and and everything was in good working order as of this morning. As far as I know, I cleaned out every single old entity. From what you're saying, it sounds like this should have worked, but clearly only partially worked. after upgrading to 2021.3.0b1.
> 
> Is there a fix? Or will I just have to bite the bullet eventually and re-do everything again if I want to get past 2021.3.0b0?

I'm looking into it.

raman325 commented:
I found the reason that some users were facing issues going from b0 to b1. It will be fixed as part of https://github.com/home-assistant/core/pull/47072

karl-gustav commented:
@raman325 so if I delete the new entites, and leave the orphaned entites (those that have the correct id and name) before i update to the next HA beta version, I will be golden?

raman325 commented:
You don't have to do anything special. Once you upgrade to b2, everything should clean itself up automatically. Any duplicate entities will get deleted, the rest will get migrated over

gregg098 commented:
Looks like this is fixed so far (will let the network heal and do some testing later on), but just a heads up that you should restart ZwaveJS2MQTT after you upgrade. All of my devices went to unavailable until I did this.
Thanks for the quick turnaround @raman325
